### ISSUE TITLE
Foundation for the Global Design System component library

### ISSUE THREAD

**Issue by gregwhitworth** [openui/open-ui#1066]
I am actively authoring a blog post that will go beyond what I said in this [comment](https://github.com/openui/open-ui/issues/1017#issuecomment-2109117089) but I think we need to make some decisions and start moving forward with some important questions that need to be answered.

I still feel there are two different approaches and they are not mutually exclusive to enable a Global Design System:

1. Testing solution
2. Component Library

I have had discussions with a few people regarding both of the options above and I think it can make sense for us to tackle both but I think option #2 is the most pragmatic one in the short term. One of the key reasons I think this is the best path _right now_ forward is there are some large component library authors interested in up-streaming their solution (I'll allow them to speak for themselves) and it provides one of the key aspects I said on the call "I don't want to waste anyone's time so we need to ensure that we do our best to succeed".

In order to move forward with a component library I've put together the following list of principles and aim for us to align on the foundational component library that meets this and begin working asap on working model; site integration, etc.

**Component library principles**

1. Ships as web components
     a. This  does not preclude Open UI from including a library that help facilitate abstractions, utility methods or polyfills 
3. Meets W3C [Intellectual Rights](https://www.w3.org/copyright/intellectual-rights/)
4. Run as an Open Source solution
      a. This does NOT mean that there will not be consensus driven resolutions as key topics will be brought to the Open UI CG telecon for resolution but many of the capabilities should be able to land within a PR with asynchronous review.
6. Align implementations with Open UI global design system blueprints
7. Adopts Open UI primitives as they become available for increased adoption and feedback
8. Has decent adoption and web developer/designer community support (monetary funding is a positive but not required)

I will kick off this discussion on Jun 27th with the aim of us resolving on the foundational framework that adheres to the above principles (please recommend more if I'm missing any). If you have a component library that you recommend or you run/maintain that you would like to offer up as this foundation to bootstrap this library, then please let us know in the comments.

If you're interested in the testing approach, please comment on that as well but that will need to likewise adhere to the same principles above, but it should also adhere to the following:

**Testing Framework Principles**

1. Needs to be framework agnostic
2. Everything in _component library principles_ section

## Desired resolution:

**Proposed Resolution:** Open UI will build a component library initially built upon <component library name> and will be augmented to align with the Open UI charter and workstream goals

**Comment by claviska** [comment]
We've discussed volunteering some components we've built for Shoelace and/or the forthcoming Web Awesome (open source) project, which are a bit opinionated but do try to align with the platform when possible. Between that, my other open source projects, and possibly some of @KonnorRogers' open source projects (he's expressed a possible interest in contributing as well), we'd have a pretty good foundation to start from.

**Comment by KonnorRogers** [comment]
👋 happy to help in any way I can. I maintain Shoelace / Web Awesome with Cory. I have built a number of other components as well including recently a combobox that has given me many gray hairs trying to get VoiceOver / NVDA to read + navigate it correctly.

testing solution:

I have a library I maintain called "shadow-DOM-testing-library" that builds on top of DOM-testing-library to provide queries into the shadow dom.

when we say "framework agnostic", I'm assuming you mean test runner agnostic?

so like needs to work across Playwright, Puppeteer et al? Or do you mean web component framework agnostic?

happy to chat testing somewhere else its more relevant! Don't want to detract from the topic at hand.

**Comment by gfellerph** [comment]
@KonnorRogers my understanding of framework agnostic testing solution would be that no matter what framework the components under test are built with, they can be tested. So the testing solution should not be usable with React, Angular, lit, ??? components only for example. What benefits would a test runner agnostic testing solution have?

**Comment by KonnorRogers** [comment]
> @KonnorRogers my understanding of framework agnostic testing solution would be that no matter what framework the components under test are built with, they can be tested. So the testing solution should not be usable with React, Angular, lit, ??? components only for example. What benefits would a test runner agnostic testing solution have?

I assumed the same, the testing solution should be usable by any web component library or framework, but wanted to make sure intent was clear of "framework agnostic". 

As for being "test runner agnostic", the main benefit is that users can continue to use their "testing framework" of choice. IE: Jest, Playwright, Web Test Runner.

The problem of course with being "test framework agnostic" is more powerful runners / environments like Playwright / Puppeteer have access to a real browser and can recreate native browser behaviors like typing, clicking, accessibility trees, etc without emulating on the client.

**Comment by gregwhitworth** [comment]
Based on the last meeting, @bradfrost and I met up briefly to focus on the "what this is and what it isn't" request made by @mfreed7. Here is an early draft that we'll ultimately bring in as a PR and land on the main page of the sibling site:

------------------

The Open UI Global Design System is a combination of blueprints and offers a reference implementation of these blueprints in the form of a component library. The Global Design System has been produced through collaboration from experts across the industry to incubate the best approaches for common UI paradigms that provide a great user experience. Aspects of the Global Design System may graduate towards a proposal within Open UI for new web platform primitives or elements that compliment and improve the user and/or the developer experience.

**What it is**
* An open-source library of consumable Web Components that serve as the implementation of Open UI’s blueprints.
* Consumable - deliver components and all the best practices baked into them so developers don’t have to reinvent them.
* Vetted by experts for accessibility, internationalization, security, and privacy
* Open source - open to contributions, forking, branching, etc. 
* Themeable
* Composable and extensible
* Intuitive to use, providing developers with a cohesive & consistent API language, sensible structure, and simple syntax.
* Interoperable to be able to power any website or app.

**What it isn’t**
* The end-all-be-all of component libraries
* A substitute for standards track 
* Just another library like Bootstrap, Material Design, Ant, et al. 
* Imposing a specific aesthetic
* Just a specification or documentation - these are actually consumable, usable components!
* Comprehensive components - Just the common components, not every component in existence
* Comprehensive features -  We're aiming for 80% of use cases, not 100%.


If you’re wanting to understand the potential new elements or primitives being incubated in Open UI into the web platform workstream which can be found [here](https://open-ui.org/).

[We’ve put together a diagram that demonstrates the relationship between the Global Design System and the Platform tracks within Open UI](https://www.figma.com/board/fyhySRUS8SFoTcOlA9RonR/Global-Design-System-Workflow?node-id=0-1&t=BymjVW8KtdzHxP54-0).

## Roadmap
Not necessarily sequential

* Green light a base component library
* GDS Site is added with clear communication on the fact that it isn’t ready and stipulates what is and isn’t with a goal alpha date. 
* Working Model, component and OSS principles defined
* Identify component list to dig into for alpha release
   * Define what it means to “dig into a component” to get it ready for implementation
* Itemize components and assign to volunteers
* Begin implementation and what is means to have a completed implementation
* Test framework, CI/CD, etc
* Aim for Alpha release in mid 2025

**Comment by gregwhitworth** [comment]
@MrHBS yep, gave an update on current progress in the most recent telecon, minutes can be found [here](https://www.w3.org/2024/11/21-openui-minutes.html).
